File: algorithms/cpp/1207.unique-number-of-occurrences.cpp
/*
 * @lc app=leetcode id=1207 lang=cpp
 *
 * [1207] Unique Number of Occurrences
 *
 * Given an array of integers arr, return true if the number of occurrences of
 * each value in the array is unique, or false otherwise.
 */

// @lc code=start

#include <vector>
#include <unordered_set>
#include <unordered_map>

class Solution
{
public:
	bool uniqueOccurrences(const std::vector<int32_t> &array)
	{
		std::unordered_map<int32_t, int32_t> hm;
		for (int32_t val : array) { hm[val]++; }

		std::unordered_set<int32_t> ht;
		for (auto it = hm.begin(); it != hm.end(); it++)
		{
			auto [val,count] = *it;
			if (ht.find(count) != ht.end()) { return false; }
			ht.insert(count);
		}

		return true;
	}
};

// @lc code=end

File: algorithms/python/1207.unique-number-of-occurrences.py
#
# @lc app=leetcode id=1207 lang=python3
#
# [1207] Unique Number of Occurrences
#
# Given an array of integers arr, return true if the number of occurrences of
# each value in the array is unique or false otherwise.
#
# Constraints:
# • 1 <= arr.length <= 1000
# • -1000 <= arr[i] <= 1000
#

# @lc code=start

from collections import Counter

class Solution:
    def uniqueOccurrences(self, arr: list[int]) -> bool:

        # Complexity:
        #   • Time: O(n)
        #   • Space: O(n)

        ht = set()

        count = Counter(arr)
        for c in count.values():

            if c in ht:
                return False

            ht.add(c)

        return True

# @lc code=end

